lib/vector-core/src/event/metric/value.rs

```diff
@@ -597,14 +600,49 @@ impl ByteSizeOf for Sample {
     }
 }
 
+/// Custom serialization function which converts special `f64` values to strings.
+/// Non-special values are serialized as numbers.
+#[allow(clippy::trivially_copy_pass_by_ref)]
+fn serialize_f64<S>(value: &f64, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    if value.is_infinite() || value.is_nan() {
+        serializer.serialize_str(&format!("{value}"))
+    } else {
+        serializer.serialize_f64(*value)
+    }
+}
+
+/// Custom deserialization function for handling special f64 values.
+fn deserialize_f64<'de, D>(deserializer: D) -> Result<f64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let value: serde_json::Value = Deserialize::deserialize(deserializer)?;
```

Comment:
This too will end up allocating a string for the special values. It's a little more complicated, but you could use the [pattern here](https://stackoverflow.com/a/43627388) to implement a custom visitor that handles strings and numbers separately.

Reply:
Neat! 

Given the nature of this code, it's worth going with the more complex/optimized implementation. Updated in https://github.com/vectordotdev/vector/pull/18650/commits/342194948ccb9308cff3f1f69cce9e7bc7b4983b.